### ISSUE TITLE
Output abnormal when I applying MXNet.cpp to implement Multi-label task

### ISSUE THREAD

**Issue by realzheng** [dmlc/MXNet.cpp#31]
I use MXNet.cpp(@hjk41 ) to implement Multi-label(4 labels,33 classes per label)task,but the output is always the same【0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 **1** 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0】 whenever I change the input.
The define net structure：

```
        .........

                Symbol fc2 = FullyConnected("fc2", relu4, fc2_w, fc2_b, 256);
        Symbol relu5 = Activation("relu5", fc2, ActivationActType::relu);
        //Symbol drop2 = Dropout("drop2",relu5,0.3);

        Symbol fc31 = FullyConnected("fc31", relu5, fc31_w, fc31_b, 33);
        Symbol fc32 = FullyConnected("fc32", relu5, fc32_w, fc32_b, 33);
        Symbol fc33 = FullyConnected("fc33", relu5, fc33_w, fc33_b, 33);
        Symbol fc34 = FullyConnected("fc34", relu5, fc34_w, fc34_b, 33);

        vector<Symbol> fcVector;
        fcVector.push_back(fc31);
        fcVector.push_back(fc32);
        fcVector.push_back(fc33);
        fcVector.push_back(fc34);

        Symbol conc = Concat("concat",fcVector, 4, 1);

        Symbol captchaNet = SoftmaxOutput("softmax", conc, data_label); 
```

Is there any errors to the defined net structure?

**Comment by zhangchen-qinyinghua** [comment]
I don't know how you transform the multi-label problem. You concat results from 4 fc-layers into one and do softmax on it, what's the point?
I think you can try `Symbol::Group()` to make multi-output.
Maybe you can give your full codes here.
We don't have a multi-label classification example. So it will be nice if you can make a PR on that once you successfully implement the task.

**Comment by realzheng** [comment]
@zhangchen-qinyinghua thanks for your reply，My code is refer to [https://gist.github.com/xlvector/6923ef145e59de44ed06f21228f2f879](url) which run success in python. And the full c++ code is [here](https://github.com/realzheng/mxnet_test/blob/master/multi_label_infer.cpp)

**Comment by hjk41** [comment]
Could you check whether the output of the previous layers change with regard to different inputs? It is hard to identify the problem just by looking at the code.

**Comment by realzheng** [comment]
@hjk41 ok,but how can I get the inner layers' output? Do I need to call the initial mxnet's c++ interface? And I have another doubt about the interface implemented in the file [op.h](https://github.com/dmlc/MXNet.cpp/blob/master/include/mxnet-cpp/op.h):
`inline Symbol Concat(const std::string& symbol_name,
                     const std::vector<Symbol>& data,
                     int num_args,
                     int dim = 1) {
  return Operator("Concat")
           .SetParam("num_args", num_args)
           .SetParam("dim", dim)
(data)
           .CreateSymbol(symbol_name);
}
`
What's " .SetParam("dim", dim)(data)" represent for?

**Comment by hjk41** [comment]
1. You can bind the intermediate symbols like fc31 and then feed input to get fc31's output value.
2. That is the Operator class interface. Operator class has SetParam() methods and operator(). You can take as look at Operator.h if you are interested.

**Comment by zhangchen-qinyinghua** [comment]
@realzheng 
1.
I check and run your code. 
Since I don't have file  ```multitask-0800.params```, I just leave the parameters randomly given.
It gives a result just as expected, 4*33 vectors, which has 4*33-1 0s, and one 1s.
I also do what @hjk41 said, the output of ```fc31``` and ```conc``` are all good.

2.
I think the main problem in your code is https://github.com/realzheng/mxnet_test/blob/master/multi_label_infer.cpp#L111.
You should use:
```
SimpleBind(ctx_cpu, args_map, map<string, NDArray>(), map<string, OpReqType>(), aux_map);
```
which also feeds the ```aux_map``` to the net.

3.
Still, I didn't get why this is transform of multi-label task.

**Comment by realzheng** [comment]
@zhangchen-qinyinghua @hjk41 thanks for your suggestions.
1. I have tried :
`SimpleBind(ctx_cpu, args_map, map<string, NDArray>(), map<string, OpReqType>(), aux_map);`
before, and I found that there is no parameter in aux_map by loading file multitask-0800.params. The output is also the same as yours.
2. We have 4 labels,so I concatenate the four full-connection layers' output , I don't know if this way  make sense, but it really works in python.

**Comment by zhangchen-qinyinghua** [comment]
I deleted my last comment, because I tried the code of @realzheng , and just got the same result.
And I didn't find big problem in @realzheng 's code.
The output of concat layer changes when input changes, but the softmax result remain the same.
I am amazed to see that the randomly given parameters(gaussian N(0, 1) sampling) seem fully control the final output...

By the way, @realzheng 
You're right that there's no aux_maps in this kind of symbols.

**Comment by conopt** [comment]
The data fed to SoftmaxOutput should have shape(batch_size, num_label). and the output is the same, where output[i*num_label+j] indicates the posibility of the ith image belonging to label j.

I also checked your code, your print the variable "p" but I don't see it elsewhere.

**Comment by realzheng** [comment]
@lx75249 
1. How can I set the shape info for SoftmaxOutput, the defined interface of SoftmaxOutput in [op.h](https://github.com/dmlc/MXNet.cpp/blob/master/include/mxnet-cpp/op.h) is as below:
`inline Symbol SoftmaxOutput(const std::string& symbol_name,
                            Symbol data,
                            Symbol label,
                            mx_float grad_scale = 1,
                            mx_float ignore_label = -1,
                            bool multi_output = false,
                            bool use_ignore = false,
                            SoftmaxOutputNormalization normalization = SoftmaxOutputNormalization::null)`
1. You're right, I have not defined variable 'p', but in my local code it has been defined as "float p = dptr_out[ j];", it's just an error when I uploaded the code which does not affect the result.

**Comment by conopt** [comment]
@realzheng The shape is inferred from data and arguments. I don't have multitask-0800.params, so I can't debug it. You can give it to me or dump your symbol by captchaNet.ToJSON() to help us debug.

**Comment by realzheng** [comment]
@lx75249 @zhangchen-qinyinghua 
I have uploaded the parameter file and update the code(https://github.com/realzheng/mxnet_test), thx.

**Comment by conopt** [comment]
@realzheng I've tried your code. I think the code works as expected, that is, the output of concat layer changes while input changes (I use random input), and softmax layer correctly transforms the outputs to possibilities. I think the problem is likely related to your network or parameters, rather than mxnet.cpp.

**Comment by realzheng** [comment]
@lx75249 I just rename the parameters' name,and the output changed when the input changed. The following codes show the differences before and after the modification：
Before:
        Symbol conv1_w("conv1_w"), conv1_b("conv1_b");
        Symbol conv2_w("conv2_w"), conv2_b("conv2_b");
        Symbol conv3_w("conv3_w"), conv3_b("conv3_b");
        Symbol fc1_w("fc1_w"), fc1_b("fc1_b");
        Symbol fc2_w("fc2_w"), fc2_b("fc2_b");

```
    Symbol fc31_w("fc31_w"), fc31_b("fc31_b");
    Symbol fc32_w("fc32_w"), fc32_b("fc32_b");
    Symbol fc33_w("fc33_w"), fc33_b("fc33_b");
    Symbol fc34_w("fc34_w"), fc34_b("fc34_b");
```

After:
        Symbol conv1_w("convolution0_weight"), conv1_b("convolution0_bias");
        Symbol conv2_w("convolution1_weight"), conv2_b("convolution1_bias");
        Symbol conv3_w("convolution2_weight"), conv3_b("convolution2_bias");
        Symbol fc1_w("fullyconnected0_weight"), fc1_b("fullyconnected0_bias");
        Symbol fc2_w("fullyconnected1_weight"), fc2_b("fullyconnected1_bias");

```
    Symbol fc31_w("fullyconnected2_weight"), fc31_b("fullyconnected2_bias");
    Symbol fc32_w("fullyconnected3_weight"), fc32_b("fullyconnected3_bias");
    Symbol fc33_w("fullyconnected4_weight"), fc33_b("fullyconnected4_bias");
    Symbol fc34_w("fullyconnected5_weight"), fc34_b("fullyconnected5_bias");
```

**Comment by zhangchen-qinyinghua** [comment]
@realzheng 
So if you are using a pre-trained model, it will be a better idea to use the file containing parameters together with the net definition JSON file, as following:
https://github.com/dmlc/MXNet.cpp/blob/master/example/feature_extract/feature_extract.cpp#L43

**Comment by realzheng** [comment]
@zhangchen-qinyinghua I have tried to load the net definition file before，but it threw out Segmentation fault. I trained the model in python.

**Comment by conopt** [comment]
@realzheng After you correct parameter names, does the result meet your expectation? If so, we can close the issue since it's not due to mxnet.cpp.

**Comment by realzheng** [comment]
@lx75249 The output is still not the desired result. BTW, Does the parameter "dim" of the function [Concat](https://github.com/dmlc/MXNet.cpp/blob/master/include/mxnet-cpp/op.h#L170) have the same meaning as it in mxnet.symbol.Concat(_args, *_kwargs)---dim (int, optional, default='1') – the dimension to be concated.

**Comment by conopt** [comment]
@realzheng I have tested the operator you used (`Concat`, `FullyConnected`, `Convolution`, see https://github.com/lx75249/MxNet.cpp/blob/qa/windows/vs/QA/deepqa.hpp#L364), and they performed well. I found that the network definition in your c++ code is slightly different with the python version, I think you should try an identical network first or train the parameters yourself.

**Comment by realzheng** [comment]
@lx75249 Thanks for your suggestions, my C++ version code is transformed from the python version(https://github.com/realzheng/mxnet_test/blob/master/infer.py) which works fine(The model is trained with my own data) , and I think the only difference between them is the input parameter "dim" of Concat(https://github.com/realzheng/mxnet_test/blob/master/infer.py#L44). I'll check my code carefully again , if there is still no other solution , do I need to close the issue?

**Comment by conopt** [comment]
@realzheng For your previous question, yes, the parameter `dim` has the same meaning across all mxnet packages, and starts from 0.

**Comment by zhangchen-qinyinghua** [comment]
@realzheng 
Could you please also provide the JSON net definition file?

**Comment by realzheng** [comment]
@zhangchen-qinyinghua The net definition file is [here](https://github.com/realzheng/mxnet_test/blob/master/multitask-symbol.json) which was generated by the training code in python. And the mainly difference is the parameter [“dim”](https://github.com/realzheng/mxnet_test/blob/master/multitask-symbol.json#L342)，in python it works fine if set it as ‘0’.  In c++, setting the parameter as '0' or '1' only affects the output number(1 label or 4 labels), but there results are wrong.

**Comment by zhangchen-qinyinghua** [comment]
@realzheng 
There seems to be some thing wrong with the JSON file you uploaded.
I generate another JSON file with `infer.py` you upload, using:

```
get_ocrnet().save("./mult_net.json")
```

And I think you should try the folowing:

```
      auto captchaNet = Symbol::Load("./mult_net.json");
      int batch_size = 1;
      args_map["data"] = NDArray(Shape(batch_size, 3, 32, 90), ctx_cpu, false);//shape(batchsize,channel,width,height)
      args_map["data"].SyncCopyFromCPU(vec);
      NDArray::WaitAll();

      map<string, NDArray> paramters;
      NDArray::Load("./mxnet_test/multitask-0990.params", NULL, &paramters);

      for (const auto &k : paramters) {
        if (k.first.substr(0, 4) == "aux:") {
          auto name = k.first.substr(4, k.first.size() - 4);
          aux_map[name] = k.second.Copy(ctx_cpu);
        }
        if (k.first.substr(0, 4) == "arg:") {
          auto name = k.first.substr(4, k.first.size() - 4);
          args_map[name] = k.second.Copy(ctx_cpu);                                                                                                                                                                                    
        }
      }
      NDArray::WaitAll();
      Executor *exe = captchaNet.SimpleBind(ctx_cpu, args_map);
      exe->Forward(false);
```

**Comment by realzheng** [comment]
@zhangchen-qinyinghua Thanks,  you are right. I can get the desired results after correcting the shape parameter as `args_map["data"] = NDArray(Shape(batch_size, 3, 32, 90), ctx_cpu, false);` . But it still threw out  a "Segmentation fault" after `delete exe;` sometimes. And I'm also trying to use `Group` for multi-task now, it works too, but it also have a seg fault sometimes. BTW, are there other  solutions to replace `SyncCopyFromCPU` , `NDArray out_cpu = out[i].Copy(ctx_cpu)` ,because `NDArray::WaitAll();` is very time consuming.

**Comment by conopt** [comment]
@realzheng You can remove `NDArray::WaitAll()` unless you really need the data (like `GetData()`).

It seems the output abnormality has been resolved. Close this issue for now. As for the segmentation fault, you can open another issue.

**Comment by WANG-MengJiao** [comment]
@lx75249 I am using a c++ api for mxnet to extract a 512-dim feature vector from images. The feature vector will be saved as .xml file. I also encounter with the problem that with/without `NDArray::WaitAll()` generate totally different outputs. Among the results, with `NDArray::WaitAll()` can get correct vector. So I am wondering if I remove the `NDArray::WaitAll()`, how can I get the correct output. Is there any trick I should do?

**Comment by conopt** [comment]
@WANG-MengJiao This repository has been moved to https://github.com/apache/incubator-mxnet/tree/master/cpp-package 
If you are using the latest version, please submit an issue in that repository. Thanks!

**Comment by konioyxgq** [comment]
@WANG-MengJiao  @lx75249  I have the same problem, have you solved it? Is there any trick you do?
